### PR TITLE
RecordTestUtils and KafkaAssertions

### PR DESCRIPTION
**Pull request by tombentley**
### Type of change

- Refactoring

### Description

Previously we were using `TestingRecord` to provide records for tests. This had the limitation that it only supports records with values and headers. It also didn't support methods like `sizeInBytes`. This could lead to subtle problems when writing tests. This PR removes `TestingRecord` and replaces it with some factory methods on `RecordTestUtils` which return "real" records, just as ~nature~Kafka intended.

It also adds some assertj assertions for Record, RecordBatch etc. These are not used in this PR, but will be in a subsequent PR.

### Additional Context

_Why are you making this pull request?_

### Checklist

_Please go through this checklist and make sure all applicable tasks have been done_

- [ ] Write tests
- [ ] Make sure all tests pass
- [ ] Review performance test results. Ensure that any degradations to performance numbers are understood and justified.
- [ ] Make sure all Sonar-Lint warnings are addressed or are justifiably ignored.
- [ ] Update documentation
- [ ] Reference relevant issue(s) and close them after merging
- [ ] For user facing changes, update CHANGELOG.md (remember to include changes affecting the API of the test artefacts too).

### CODE REVIEW

File: kroxylicious-filter-test-support/src/main/java/io/kroxylicious/test/record/RecordTestUtils.java [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+                ByteBuffer.wrap(key.getBytes(StandardCharsets.UTF_8)),
+                ByteBuffer.wrap(value.getBytes(StandardCharsets.UTF_8)),
+                headers);
+    }
+
+    public static MemoryRecords memoryRecords(byte magic, long offset, long timestamp, ByteBuffer key, ByteBuffer value, Header... headers) {
+        MemoryRecordsBuilder memoryRecordsBuilder = new MemoryRecordsBuilder(
+                ByteBuffer.allocate(1024),
+                magic,
+                CompressionType.NONE,
```
Comment (robobario) [inline kroxylicious-filter-test-support/src/main/java/io/kroxylicious/test/record/RecordTestUtils.java:94]: Could be worth adding another method with all the batch configs exposed?

**Comment by robobario** [comment]
> This could lead to subtle problems when writing tests

Agreed, the real thing is subtle so better to use it.

Also not sure what's happened to the github builds

File: kroxylicious-filter-test-support/src/main/java/io/kroxylicious/test/record/RecordTestUtils.java [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+
+import org.apache.kafka.common.header.Header;
+import org.apache.kafka.common.record.CompressionType;
+import org.apache.kafka.common.record.MemoryRecords;
+import org.apache.kafka.common.record.MemoryRecordsBuilder;
+import org.apache.kafka.common.record.Record;
+import org.apache.kafka.common.record.RecordBatch;
+import org.apache.kafka.common.record.TimestampType;
+
+public class RecordTestUtils {
```
Comment (robobario) [inline kroxylicious-filter-test-support/src/main/java/io/kroxylicious/test/record/RecordTestUtils.java:20]: Did you have some other cases in mind beyond factory methods? Maybe `RecordFactory` describes it better at the moment, or we could use `Records`.

File: kroxylicious-filters/kroxylicious-encryption/src/test/java/io/kroxylicious/filter/encryption/EnvelopeEncryptionFilterTest.java [hunk 3]
```diff
@@ -186,7 +186,7 @@ void shouldDecryptEncryptedRecords() {
 
         when(keyManager.decrypt(any(), anyInt(), assertArg(records -> assertThat(records).hasSize(1)), any(Receiver.class))).thenAnswer(invocationOnMock -> {
             final Receiver receiver = invocationOnMock.getArgument(3);
-            receiver.accept(new TestingRecord(plainTextPayload.messageBytes()), plainTextPayload.messageBytes(), Record.EMPTY_HEADERS);
+            receiver.accept(RecordTestUtils.record(plainTextPayload.messageBytes()), plainTextPayload.messageBytes(), Record.EMPTY_HEADERS);
```
Comment (robobario) [inline kroxylicious-filters/kroxylicious-encryption/src/test/java/io/kroxylicious/filter/encryption/EnvelopeEncryptionFilterTest.java:189]: I could imagine the no-header case being pretty common for testing, maybe we should have a convenience like `RecordTestUtils.record(keyBytes, valueBytes)` .

**Review by robobario — APPROVED** [review]
LGTM

File: kroxylicious-filter-test-support/src/main/java/io/kroxylicious/test/record/RecordTestUtils.java (hunk 1 above)
Comment (tombentley) [reply kroxylicious-filter-test-support/src/main/java/io/kroxylicious/test/record/RecordTestUtils.java:94]: Easily done in a PR where such a method is needed.

File: kroxylicious-filters/kroxylicious-encryption/src/test/java/io/kroxylicious/filter/encryption/EnvelopeEncryptionFilterTest.java (hunk 3 above)
Comment (tombentley) [reply kroxylicious-filters/kroxylicious-encryption/src/test/java/io/kroxylicious/filter/encryption/EnvelopeEncryptionFilterTest.java:189]: This `Record.EMPTY_HEADERS` is actually being passed as an argument to `Receiver`, not `RecordTestUtils.record()`

File: kroxylicious-filter-test-support/src/main/java/io/kroxylicious/test/record/RecordTestUtils.java (hunk 2 above)
Comment (tombentley) [reply kroxylicious-filter-test-support/src/main/java/io/kroxylicious/test/record/RecordTestUtils.java:20]: Not really. Right now the intention is that these methods are just for tests. Class names like `RecordFactory` make it less obvious that it's not really a general purpose thing. So I'd rather keep the ugly name.

**Pull request merged**